### PR TITLE
FEAT: Identify coupling 5649

### LINKED ISSUE #5649: Add a function which will identify coupling in a specific range - plot and report identified coupling in a file
### Description of the feature

This feature is useful for large RF board where user want to identified unwanted coupling.
User will specify:

- Sparameter file full path
- Log file full path
- low_loss and high_loss range
- Frequency start at which to start looking for the coupling(such we can skip low frequency where DC block exist)
- Sample frequency(to speed up process chack coupling at each sample)

### Steps for implementing the feature

We will  use TouchstoneData to create this function

### Useful links and references

_No response_

### PR DESCRIPTION
**Pull request by amichel0205**
## Description

This feature is useful for large RF board where user want to identified unwanted coupling.
User will specify:

Sparameter file full path
Log file full path
low_loss and high_loss range
Frequency start at which to start looking for the coupling(such we can skip low frequency where DC block exist)
Sample frequency(to speed up process chack coupling at each sample)

## Issue linked
Close #5649 

## Checklist
- [x] I have tested my changes locally.
- [x] I have added necessary documentation or updated existing documentation.
- [x] I have followed the coding style guidelines of this project.
- [x] I have added appropriate tests (unit, integration, system).
- [x] I have reviewed my changes before submitting this pull request.
- [x] I have linked the issue or issues that are solved by the PR if any.
- [x] I have agreed with the Contributor License Agreement ([CLA](https://developer.ansys.com/form/cla-acceptance)).

### CODE REVIEW

**Review by hui-zhou-a — COMMENTED** [review]
@amichel0205 please use a valid s-parameter for testing.

File: tests/system/general/test_44_TouchstoneParser.py [hunk 1]
```diff
@@ -119,6 +119,13 @@ def test_get_next_xtalk_index_with_dummy_prefix(touchstone_file):
     assert not res
 
 
+def test_get_coupling_in_range(touchstone_file):
+    ts = TouchstoneData(touchstone_file=touchstone_file)
+    res = ts.get_coupling_in_range(start_at_frequency=1e9, high_loss=-60, low_loss=-40, frequency_sample=5, plot=True)
```
[the remark this answers is not in the document]
Comment (amichel0205) [reply tests/system/general/test_44_TouchstoneParser.py:124]: Probably not but do you think I just need to change the test to start are 0.1GHZ or should i used another S aparameter?

[the remark this answers is not in the document]
Comment (amichel0205) [reply tests/system/general/test_44_TouchstoneParser.py]: def test_get_coupling_in_range(touchstone_file):
    ts = TouchstoneData(touchstone_file=touchstone_file)
    res = ts.get_coupling_in_range(start_at_frequency=0.1e9, high_loss=-60, low_loss=-40, frequency_sample=5, plot=True)

    assert not res

I chqnged qnd push

**Comment by Samuelopez-ansys** [comment]
Hi @amichel0205 ,

I think the argument log_file_name should be the last one and not the first one, because it is only used when you want to save it. 

In addition, you should change the name, to **output_file**, we are trying to make consisten name of variables in all PyAEDT, the methods, the description could be something like:

Output file path to save .... . If ``None`` no file is saved.

The same for start_at_frequency, should be start_frequency keeping consistency.

**Review by Samuelopez-ansys — CHANGES_REQUESTED** [review]
Hi @amichel0205 ,

I think the argument log_file_name should be the last one and not the first one, because it is only used when you want to save it.

In addition, you should change the name, to output_file, we are trying to make consisten name of variables in all PyAEDT, the methods, the description could be something like:

Output file path to save .... . If None no file is saved.

The same for start_at_frequency, should be start_frequency keeping consistency.

**Review by Samuelopez-ansys — DISMISSED** [review]
LGTM

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 2]
```diff
@@ -111,6 +112,85 @@ def __init__(self, solution_data=None, touchstone_file=None):
...
+                            freq_f = freq.format((self.f[k]) * 1e-9)
+                            temp_file.append(sxy + str_loss_f + freq_f + "\n")
+                            k = k + frequency_sample
+                            break
+                        else:
+                            k = k + frequency_sample
+                            continue
+                j = j + 1
+            i = i + 1
+            j = i
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:184]:
```suggestion
for i in range(nb_port):
    for j in range(nb_port):
        if i == j:
            continue
        for k in range(k_start, nb_freq, frequency_sample):
            loss = s_db[k, i, j]
            if high_loss < loss < low_loss:
                temp_list.append((i, j))
                sxy = f"S({self.port_names[i]} , {self.port_names[j]})"
                line = f"{sxy} Loss = {loss:.2f} dB Freq = {(self.f[k] * 1e-9):.3f} GHz\n"
                temp_file.append(line)
                break
```
Btw, I was confused by `low_loss` and `high_loss` values at first. I had to do some research to understand it xD

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 3]
```diff
@@ -120,12 +200,12 @@ def get_insertion_loss_index(self, threshold=-3):
         Parameters
         ----------
         threshold : float, int, optional
-            Threshold to determine shorted ports in dB. Default value is ``3``.
+            Threshold to determine shorted ports in dB. The default is ``-3``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:203]:
```suggestion
            Threshold to determine shorted ports in dB. The default value is ``-3``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 4]
```diff
@@ -171,9 +251,9 @@ def plot(self, index_couples=None, show=True):
         Parameters
         ----------
         index_couples : list, optional
-            List of indexes couple to plot. Default is ``None`` to plot all ``port_tuples``.
+            List of indexes couple to plot. The default is ``None`` to plot all ``port_tuples``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:254]:
```suggestion
            List of indexes couple to plot. The default value is ``None`` to plot all ``port_tuples``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 5]
```diff
@@ -171,9 +251,9 @@ def plot(self, index_couples=None, show=True):
...
         show : bool
-            Whether to plot. Default is ``True``.
+            Whether to plot. The default is ``True``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:256]:
```suggestion
            Whether to plot. The default value is ``True``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 6]
```diff
@@ -260,7 +340,7 @@ def get_return_loss_index(self, excitation_name_prefix=""):
         Parameters
         ----------
         excitation_name_prefix :str, optional
-            Prefix of the excitation. Default value is ``""``.
+            Prefix of the excitation. The default is ``""``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:343]:
```suggestion
            Prefix of the excitation. The default value is ``""``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 7]
```diff
@@ -314,7 +394,7 @@ def get_next_xtalk_index(self, tx_prefix=""):
         Parameters
         ----------
         tx_prefix :str, optional
-            Prefix for TX (eg. "DIE"). Default value is ``""``.
+            Prefix for TX (eg. "DIE"). The default is ``""``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:397]:
```suggestion
            Prefix for TX (eg. "DIE"). The default value is ``""``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 8]
```diff
@@ -347,7 +427,7 @@ def get_fext_xtalk_index_from_prefix(self, tx_prefix, rx_prefix, skip_same_index
         rx_prefix : str
             prefix for RX (eg. "BGA")
         skip_same_index_couples : bool
-            Boolean ignore TX and RX couple with same index. The default value is ``True``.
+            Boolean ignore TX and RX couple with same index. The default is ``True``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:430]:
```suggestion
            Boolean ignore TX and RX couple with same index. The default value is ``True``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 9]
```diff
@@ -369,7 +449,7 @@ def plot_next_xtalk_losses(self, tx_prefix=""):
         Parameters
         ----------
         tx_prefix: str, optional
-            Prefix for TX. Default value is ``""``.
+            Prefix for TX. The default is ``""``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:452]:
```suggestion
            Prefix for TX. The default value is ``""``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 10]
```diff
@@ -415,15 +495,16 @@ def get_worst_curve(self, freq_min=None, freq_max=None, worst_is_higher=True, cu
         Parameters
         ----------
         freq_min : float, optional
-            Minimum frequency to analyze in GHz (None to 0). Default value is ``None``.
+            Minimum frequency to analyze in GHz (None to 0). The default is ``None``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:498]:
```suggestion
            Minimum frequency to analyze in GHz (None to 0). The default value is ``None``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 11]
```diff
@@ -415,15 +495,16 @@ def get_worst_curve(self, freq_min=None, freq_max=None, worst_is_higher=True, cu
...
         worst_is_higher : bool
-            Worst curve is the one with higher mean value. Default value is ``True``.
+            Worst curve is the one with higher mean value. The default value is ``True``.
         curve_list : list
-            List of [m,n] index of curves on which to search. None to search on all curves. Default value is ``None``.
+            List of [m,n] index of curves on which to search. None to search on all curves.
+            The default value is ``None``.
         plot : bool, optional
-            Whether to plot or not the chart.
+            Whether to plot or not the chart. The default is ``True``.
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:507]:
```suggestion
            Whether to plot or not the chart. The default value is ``True``.
```

File: tests/system/general/test_44_TouchstoneParser.py [hunk 12]
```diff
@@ -81,6 +81,16 @@ def test_03_check_touchstone_file(self):
...
 
+    def test_get_coupling_in_range(self, local_scratch):
+        touchstone_file = os.path.join(test_T44_dir, "port_order_1234.s8p")
+        output_file = os.path.join(self.local_scratch.path, "test_44_gcir.log")
+        ts = TouchstoneData(touchstone_file=touchstone_file)
+        res = ts.get_coupling_in_range(
+            start_frequency=1e9, high_loss=-60, low_loss=-40, frequency_sample=5, output_file=output_file
+        )
+
+        assert res
```
Comment (SMoraisAnsys) [inline tests/system/general/test_44_TouchstoneParser.py:92]: Please check that `res` is equal to a list with some value and not only that it's not "`False` or `None`"

**Review by SMoraisAnsys — CHANGES_REQUESTED** [review]
I left some refactoring proposition.

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py (hunk 2 above)
Comment (amichel0205) [reply src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:184]: Loss is negative in dB so as an example low_loss is -20dB as high_loss is -40dB.

Comment (amichel0205) [reply src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:184]: HI @SMoraisAnsys 

I check the results comparing your suggestion to the original code for the log file and the results is different so I've to check why. Your suggestion identify more candidate for the test "if high_loss < loss < low_loss:"

I would commit your change as it is much cleaner but I need to understand before what is the difference.

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 13]
```diff
@@ -111,6 +112,85 @@ def __init__(self, solution_data=None, touchstone_file=None):
...
+                            k = k + frequency_sample
+                            break
+                        else:
+                            k = k + frequency_sample
+                            continue
+                j = j + 1
+            i = i + 1
+            j = i
+        if output_file is not None:
+            logger.info("if file " + output_file + " exist, the file will be overwritten.")
```
[the remark this answers is not in the document]
Comment (amichel0205) [reply src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:186]: Suggestion implemented and will be available in next commit.

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py (hunk 3 above)
Comment (amichel0205) [reply src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:203]: This is not my code but I will commit your suggestion.

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py (hunk 4 above)
Comment (amichel0205) [reply src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:254]: This is not my code but will commit your change.

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py (hunk 2 above)
Comment (amichel0205) [reply src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:184]: Hi @SMoraisAnsys 

I know why we have a difference because of reciprocity I check only 1 side of the matrix:
image
Your suggested code check the two side of the matrix
such you can see below that in the log file you have both S(MA4_I , MA4_O) and S(MA4_O , MA4_I)
image

I think in my code I achieve it by doing 
j = i
at the end of the while j < nb_port: loop

Do you have a suggestion to fix this?

File: tests/system/general/test_44_TouchstoneParser.py (hunk 12 above)
Comment (amichel0205) [reply tests/system/general/test_44_TouchstoneParser.py:92]: HI @SMoraisAnsys 

Replace "assert res " by "assert isinstance(res, list)"

**Review by hui-zhou-a — APPROVED** [review]
LGTM

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 14]
```diff
@@ -120,12 +188,12 @@ def get_insertion_loss_index(self, threshold=-3):
         Parameters
         ----------
         threshold : float, int, optional
-            Threshold to determine shorted ports in dB. Default value is ``3``.
+            Threshold to determine shorted ports in dB. The default is ``-3``.
```
Comment (Samuelopez-ansys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:191]:
```suggestion
            Threshold to determine shorted ports in dB. The default value is ``-3``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 15]
```diff
@@ -171,9 +239,9 @@ def plot(self, index_couples=None, show=True):
         Parameters
         ----------
         index_couples : list, optional
-            List of indexes couple to plot. Default is ``None`` to plot all ``port_tuples``.
+            List of indexes couple to plot. The default is ``None`` to plot all ``port_tuples``.
```
Comment (Samuelopez-ansys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:242]:
```suggestion
            List of indexes couple to plot. The default value is ``None`` to plot all ``port_tuples``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 16]
```diff
@@ -347,7 +415,7 @@ def get_fext_xtalk_index_from_prefix(self, tx_prefix, rx_prefix, skip_same_index
         rx_prefix : str
             prefix for RX (eg. "BGA")
         skip_same_index_couples : bool
-            Boolean ignore TX and RX couple with same index. The default value is ``True``.
+            Boolean ignore TX and RX couple with same index. The default is ``True``.
```
Comment (Samuelopez-ansys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:418]:
```suggestion
            Boolean ignore TX and RX couple with same index. The default value is ``True``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 17]
```diff
@@ -369,7 +437,7 @@ def plot_next_xtalk_losses(self, tx_prefix=""):
         Parameters
         ----------
         tx_prefix: str, optional
-            Prefix for TX. Default value is ``""``.
+            Prefix for TX. The default is ``""``.
```
Comment (Samuelopez-ansys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:440]:
```suggestion
            Prefix for TX. The default value is ``""``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 18]
```diff
@@ -415,15 +483,16 @@ def get_worst_curve(self, freq_min=None, freq_max=None, worst_is_higher=True, cu
         Parameters
         ----------
         freq_min : float, optional
-            Minimum frequency to analyze in GHz (None to 0). Default value is ``None``.
+            Minimum frequency to analyze in GHz (None to 0). The default is ``None``.
```
Comment (Samuelopez-ansys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:486]:
```suggestion
            Minimum frequency to analyze in GHz (None to 0). The default value is ``None``.
```

File: src/ansys/aedt/core/visualization/advanced/touchstone_parser.py [hunk 19]
```diff
@@ -415,15 +483,16 @@ def get_worst_curve(self, freq_min=None, freq_max=None, worst_is_higher=True, cu
...
         worst_is_higher : bool
-            Worst curve is the one with higher mean value. Default value is ``True``.
+            Worst curve is the one with higher mean value. The default value is ``True``.
         curve_list : list
-            List of [m,n] index of curves on which to search. None to search on all curves. Default value is ``None``.
+            List of [m,n] index of curves on which to search. None to search on all curves.
+            The default value is ``None``.
         plot : bool, optional
-            Whether to plot or not the chart.
+            Whether to plot or not the chart. The default is ``True``.
```
Comment (Samuelopez-ansys) [inline src/ansys/aedt/core/visualization/advanced/touchstone_parser.py:495]:
```suggestion
            Whether to plot or not the chart. The default value is ``True``.
```
